models/demos/utils/llm_demo_utils.py

```diff
@@ -0,0 +1,114 @@
+    - The targets dictionary should contain the following keys: "prefill_t/s", "decode_t/s", "decode_t/s/u"
+    - The N_warmup_iter dictionary should contain the following keys: "inference_prefill", "inference_decode"
+    """
+
+    assert all(
+        key in measurements
+        for key in [
+            "compile_prefill",
+            "compile_decode",
+            "prefill_t/s",
+            "decode_t/s",
+            "decode_t/s/u",
+        ]
+    )
+
+    benchmark_data = BenchmarkData()
+
+    # Add measurement data
+    benchmark_data.add_measurement(profiler, 0, "compile_prefill", "time(s)", measurements["compile_prefill"])
+    benchmark_data.add_measurement(profiler, 0, "compile_decode", "time(s)", measurements["compile_decode"])
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_prefill",
+        "tokens/s",
+        measurements["prefill_t/s"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_prefill"]
+        if "inference_prefill" in N_warmup_iter
+        else None,
+        target=targets["prefill_t/s"] if "prefill_t/s" in targets else None,
+    )
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_decode",
+        "tokens/s",
+        measurements["decode_t/s"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_decode"] if "inference_decode" in N_warmup_iter else None,
+        target=targets["decode_t/s"] if "decode_t/s" in targets else None,
+    )
+    benchmark_data.add_measurement(
+        profiler,
+        0,
+        "inference_decode",
+        "tokens/s/user",
+        measurements["decode_t/s/u"],
+        step_warm_up_num_iterations=N_warmup_iter["inference_decode"] if "inference_decode" in N_warmup_iter else None,
+        target=targets["decode_t/s/u"] if "decode_t/s/u" in targets else None,
+    )
+    return benchmark_data
+
+
+def verify_tokens(generated_text: dict, expected_greedy_output_path: str):
+    with open(expected_greedy_output_path, "r") as f:
+        expected_output = json.load(f)
+    does_pass = generated_text == expected_output
+    if does_pass:
+        logger.info("Output Check Passed!")
+    else:
+        assert (
+            does_pass
+        ), f"Generated text does not match expected output! \n\n Generated text:\n {generated_text} \n\n Expected output:\n {expected_output}"
+
+
+def verify_perf(measurements: dict, expected_perf_metrics: dict):
+    """
+    Verify the performance metrics against the expected values.
+    The metrics that must be provided are specified in expected_measurements below.
+    """
+
+    expected_measurements = {
+        "compile_prefill": False,
+        "compile_decode": False,
+        "prefill_t/s": True,
+        "decode_t/s": True,
+        "decode_t/s/u": True,
+    }
+
+    does_pass = True
+    for key in expected_measurements:
+        if not expected_measurements[key]:
```

Comment:
Just a small thing, but you may want to assert later that you actually did find at least one measurement. What if we go through all of them and none of them exist? Unless I'm reading this wrong, this function would pass

Unless your intended behaviour in that case is to pass.

Reply:
If all of the entries in expected_measurements are False (no targets are expected), it is intended to pass. The ones that are specified as True are currently asserted against and so it will fail if they are not in the measurements/expected_perf_metrics.